### ISSUE TITLE
[BUG] - cardano-cli protocol-state CandidateNonce does not match epochNonce post epoch transition

### ISSUE THREAD

**Issue by rdlrt** [input-output-hk/cardano-node#4352]
*External*

**Area**
*Other*   Any other topic (Delegation, Ranking, ...).

**Summary**

The CandidateNonce created mins before epoch transition does not match the epochNonce of next epoch.

**Steps to reproduce**
Steps to reproduce the behavior (for example, this is on Preview network):
1. Wait for 2/3rd of epoch transition
2. Execute : `cardano-cli query protocol-state --testnet-magic 2 | jq .` and capture CandidateNonce
3. Wait for epoch transition, and then re-execute the command above and check new `epochNonce`

**Expected behavior**

The above should match?

**System info (please complete the following information):**
- OS Name: Irrelevant
- OS Version: Irrelevant
- Node version (output of `cardano-node --version`): 1.35.3, 80b5637a5520648d50b0763d7677ddbe374cd598
- CLI version  (output of `cardano-cli --version`): 1.35.3, 80b5637a5520648d50b0763d7677ddbe374cd598

**Screenshots and attachments**

Preview Network, 5 mins before epoch transition from epoch 8 to 9

``` JSON
{
  "candidateNonce": {
    "contents": "20d74b68eee33a509be02a1351543736bfcf79bc1e5570c5e3635f4d1be874d7",
    "tag": "Nonce"
  },
  "epochNonce": {
    "contents": "abb95f45898f342148df6246d844b10549a19b914b8b91a91731113733478404",
    "tag": "Nonce"
  },
  "evolvingNonce": {
    "contents": "83e38041f59424317d8a4376c4ef15576172cc9c18498cf551f04b96c47e56ed",
    "tag": "Nonce"
  },
  "labNonce": {
    "contents": "b8d3fb8244de39bcb06f584fc5b6d6f0ab7dfca2c7ebe9ef476c5d0a8bd5313a",
    "tag": "Nonce"
  },
  "lastEpochBlockNonce": {
    "contents": "1d3a1b2d31ea7fd81529d3e5fda711d837e82913733755910c7ef219db7bed4e",
    "tag": "Nonce"
  },
  "oCertCounters": {
    "...": 0,
  }
}
```
Actual Nonce in epoch 9:
``` bash
cardano-cli query protocol-state --testnet-magic 2 | jq -r .epochNonce.contents
2753132054e4b57b677399c9c8234562610efa9962c4c9429fd2bed9a81ff130
```

**Comment by dostrelith678** [comment]
It's been more than a month since this issue was opened. Could we please get an update on this from someone? I have reproduced the bug with the steps above as well.

**Comment by JaredCorduan** [comment]
@rdlrt - I don't see a bug here.

* The candidate nonce does not freeze at the 2/3 mark in the epoch, but rather at the 7/10 mark (one stability window from the end). See "Figure 60: Update Nonce rules" of [Shelley ledger spec](https://hydra.iohk.io/job/Cardano/cardano-ledger/shelleyLedgerSpec/latest/download-by-type/doc-pdf/ledger-spec).
* The candidate nonce is combined with a block header hash to produce the next epoch nonce. See "Figure 58: Tick Nonce rules" of [Shelley ledger spec](https://hydra.iohk.io/job/Cardano/cardano-ledger/shelleyLedgerSpec/latest/download-by-type/doc-pdf/ledger-spec). Note that the extra-entropy field can also change the epoch nonce, but we used that feature precisely [once](https://iohk.io/en/blog/posts/2021/03/29/the-secure-transition-to-decentralization/), and it was removed with the Vasil HF.

The spec uses the following notation:

* $η_0$ - the epoch nonce
* $η_c$ - the candidate nonce
* $η_{ph}$ - the previous hash of the first block in the epoch
* $η_c$ - the extra entropy

**Comment by rdlrt** [comment]
> The candidate nonce does not freeze at the 2/3 mark in the epoch

Indeed, that was an inaccurate generalised statement to mention 2/3rd, but the nonce we're looking at was 5 mins before end of 1-hour long epoch.

> The candidate nonce is combined with a block header hash to produce the next epoch nonce

That's interesting, thanks - was unaware that it's not usable as-is. But then also, this would mean we cannot rely on fetching epoch nonce for next epoch using output from CLI , without post-processing on client side (as against epoch nonce which can be directly parsed from output). Do you think it would be a contention open for a change?

**Comment by JaredCorduan** [comment]
> But then also, this would mean we cannot rely on fetching epoch nonce for next epoch purely using output from CLI (without post-processing on client side) - as against epoch nonce for current epoch which is available from protocol-stake readily usable. Do you think it would be a contention open for a change for CLI output?

I think it's a totally reasonable feature request for the CLI to supply an _additional_ field for the `nextEpochNonce`, which would either let the user know that the nonce is not yet frozen, or provide the value for the next epoch.

I had a quick look, and I think that `lastEpochBlockNonce` is what the ledger spec calls $η_{ph}$. This means that once the candidate nonce is frozen, the next epoch nonce will be the BLAKE2b-256 hash of the concatenation `candidateNonce|lastEpochBlockNonce` (the "seed operation" used in the ledger spec is specified in `A.6 Abstract functions`, and is implemented [here](https://github.com/input-output-hk/cardano-ledger/blob/3dae90e864f2fa51cdb7f16f6346284231f7495f/libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs#L448-L454) in the ledger code).

**Comment by JaredCorduan** [comment]
@rdlrt would you like to edit the original issue to make it clear that this is now a feature request? or alternatively, you could close this one and make a new one (which references this one). I just don't want it to be obscured (and I don't myself often work in this repo, someone else will likely have to pick it up).

**Comment by rdlrt** [comment]
> could you like to edit the original issue

done, thanks
